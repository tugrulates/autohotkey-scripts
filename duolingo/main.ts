import { Command } from "@cliffy/command";
import { Table } from "@cliffy/table";
import { pool } from "@tugrulates/internal/async";
import { Config } from "@tugrulates/internal/cli";
import { DuolingoClient } from "./client.ts";
import { LEAGUES } from "./data.ts";
import {
  engageWithCard,
  followLeagueUsers,
  getEmoji,
  getLeagueUserEmoji,
} from "./interaction.ts";
import type { FeedCard, League } from "./types.ts";

let username: Config | undefined;
let token: Config | undefined;

/** Duolingo client built from common CLI options. */
export async function getClient(): Promise<DuolingoClient> {
  if (!username || !token) {
    throw new Error("Username and token not configured.");
  }
  return new DuolingoClient(await username.get(), await token.get());
}

/**
 * Fetches and organizes follow information.
 *
 * @returns Users who are followed, users who follow, and their difference sets.
 */
async function getFollows() {
  const client = await getClient();
  const [following, followers] = await Promise.all([
    client.getFollowing(),
    client.getFollowers(),
  ]);
  return {
    following,
    followers,
    dontFollowBack: following.filter(({ userId }) =>
      !followers.some((user) => user.userId === userId)
    ),
    notFollowingBack: followers.filter(({ userId }) =>
      !following.some((user) => user.userId === userId)
    ),
  };
}

/**
 * Outputs the league to the console.
 *
 * @param league League to output.
 */
async function outputLeague(league: League) {
  const client = await getClient();
  const following = await client.getFollowing();
  const tier = LEAGUES[league.tier];
  new Table()
    .header([tier.emoji, tier.name])
    .body(
      league.rankings.map((user, index) => [
        `${index + 1}.`,
        `${user.display_name} ${getLeagueUserEmoji(user)}`,
        following ? "👤" : "",
        `${user.score.toString()} XP`,
      ]),
    )
    .columns([{ align: "right" }, {}, {}, { align: "right" }])
    .render();
}

/**
 * Returns the display summary of the card.
 *
 * @param card Card to get the summary for.
 * @returns Display summary of the card.
 */
function getSummary(card: FeedCard): string {
  return card.header?.replace(/<[^>]+>/g, "") ??
    `${card.displayName} ${card.body.toLowerCase()}`;
}

function getFeedCommand() {
  return new Command()
    .description("Prints and interacts with the feed.")
    .example("duolingo feed", "Prints the feed.")
    .example("duolingo feed --engage", "Engages with the feed.")
    .example("duolingo feed --json | jq", "Query JSON over the feed.")
    .option("--engage", "Engage with the feed events.")
    .option("--json", "Output the feed as JSON.")
    .action(async ({ engage, json }) => {
      const client = await getClient();
      const followers = await client.getFollowers();
      const cards = await client.getFeedCards();
      if (json) console.log(JSON.stringify(cards, undefined, 2));
      await pool(
        cards,
        async (card) => {
          if (!engage || await engageWithCard(client, followers, card)) {
            if (!json) console.log(`${getEmoji(card)} ${getSummary(card)}`);
          }
        },
      );
    });
}

function getFollowsCommand() {
  return new Command()
    .description("Prints and manages follower information on Duolingo.")
    .example("duolingo follows", "Prints follow counts.")
    .example("duolingo follows --follows", "Follow users who follow.")
    .example("duolingo follows --unfollow", "Unfollow users who dont' follow.")
    .example("duolingo follows --follow --unfollow", "Matches both lists.")
    .example("duolingo follows --json", "Outputs JSON of follower information.")
    .example(
      "duolingo follows --json | jq",
      "Query JSON for follower information.",
    )
    .example(
      "duolingo follows --json | jq '.dontFollowBack[].username'",
      "List users who are followed but don't follow back.",
    )
    .example(
      "duolingo follows --json | jq '.notFollowingBack[].username'",
      "List users who follow but are not followed back.",
    )
    .option("--follow", "Follow users who follow.")
    .option("--unfollow", "Unfollow users who don't follow.")
    .option("--json", "Output the follower information as JSON.")
    .action(async ({ follow, unfollow, json }) => {
      const client = await getClient();
      let result = await getFollows();

      if (follow || unfollow) {
        if (follow) {
          await pool(
            result.notFollowingBack,
            async (user) => {
              await client.followUser(user.userId);
              if (!json) console.log(`✅ Followed ${user.username}.`);
            },
          );
        }
        if (unfollow) {
          await pool(
            result.dontFollowBack,
            async (user) => {
              await client.unfollowUser(user.userId);
              if (!json) console.log(`❌ Unfollowed ${user.username}.`);
            },
          );
        }
        result = await getFollows();
      }

      if (json) console.log(JSON.stringify(result, undefined, 2));
      else {
        console.log(`👤 Following ${result.following.length} people.`);
        console.log(`👤 Followed by ${result.followers.length} people.`);
      }
    });
}

function getLeagueCommand() {
  return new Command()
    .description("Prints and interacts with the current Duolingo league.")
    .example("duolingo league", "Prints the league.")
    .example("duolingo league --follow", "Follows users in the league.")
    .example("duolingo league --json | jq", "Query JSON over the league.")
    .option("--follow", "Follow users in the league.")
    .option("--json", "Output the league as JSON.")
    .action(async ({ follow, json }) => {
      const client = await getClient();
      const league = await client.getLeague();
      if (follow) await followLeagueUsers(client, league.rankings);
      if (json) console.log(JSON.stringify(league, undefined, 2));
      else await outputLeague(league);
    });
}

async function getCommand() {
  const command = new Command()
    .name("duolingo")
    .description("Interact with Duolingo.")
    .example("duolingo --username <username> --token <token>", "Configure.")
    .example("duolingo --clear", "Clear the cached configuration.")
    .usage("<command> [options]")
    .globalOption(
      "--username <username:string>",
      "Username.",
      await username?.option(),
    )
    .globalOption(
      "--token <token:string>",
      "JWT token.",
      await token?.option(),
    )
    .option("--clear", "Clear the cached configuration.", {
      standalone: true,
      action: async () => {
        await username?.clear();
        await token?.clear();
      },
    })
    .action((): void => command.showHelp())
    .command("feed", getFeedCommand())
    .command("follows", getFollowsCommand())
    .command("league", getLeagueCommand());
  return command;
}

/** CLI entrypoint. */
export async function main() {
  username = new Config("username");
  token = new Config("token", { secret: true });
  const command = await getCommand();
  await command.parse();
}
