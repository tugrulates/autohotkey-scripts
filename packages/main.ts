import { Command } from "@cliffy/command";
import { Table } from "@cliffy/table";
import {
  bumpVersion,
  findCommitsSinceLastRelease,
  getPackage,
  getWorkspacePackages,
  setPackage,
} from "@tugrulates/internal/package";

function getListCommand() {
  return new Command()
    .description("Lists all packages in the workspace.")
    .example("packages list", "List all Deno packages.")
    .action(async () => {
      const packages = await getWorkspacePackages(".");
      new Table()
        .body(
          packages.map((
            pkg,
          ) => ["ðŸ“¦", pkg.directory, pkg.config.name, pkg.config.version]),
        )
        .render();
    });
}

function getBumpCommand() {
  return new Command()
    .description("Prints a list of active and high quality users on 500px.")
    .example("packages bump", "Updates versions for packages with a change.")
    .example("packages bump --dry-run", "Calculate updates without changes.")
    .option("--dry-run", "Calculate updates without changes.")
    .action(async ({ dryRun }) => {
      const packages = await getWorkspacePackages(".");
      const updates = (await Promise.all(
        packages.filter((pkg) => pkg.config.name && pkg.config.version).map(
          async (pkg) => {
            const commits = await findCommitsSinceLastRelease(pkg);
            const version = bumpVersion(pkg, commits);
            return {
              ...pkg,
              newVersion: version,
              commits,
            };
          },
        ),
      )).filter((pkg) => pkg.config.version !== pkg.newVersion);
      if (!dryRun) {
        await Promise.all(
          updates.map((pkg) =>
            setPackage({
              ...pkg,
              config: { ...pkg.config, version: pkg.newVersion },
            })
          ),
        );
      }
      new Table().body(
        updates.map((
          pkg,
        ) => ["ðŸ“Œ", pkg.config.name, pkg.config.version, "ðŸ‘‰", pkg.newVersion]),
      ).render();
    });
}

async function getCommand() {
  const command = new Command()
    .name("packages")
    .description("Manage repository packages.")
    .usage("<command> [options]")
    .version((await getPackage()).config.version ?? "")
    .action((): void => command.showHelp())
    .command("list", getListCommand())
    .command("bump", getBumpCommand());
  return command;
}

/** CLI entrypoint. */
async function main(args: string[]) {
  const command = await getCommand();
  await command.parse(args);
}

if (import.meta.main) await main(Deno.args);
