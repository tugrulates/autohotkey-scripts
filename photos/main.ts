import { colors } from "@cliffy/ansi/colors";
import { Command } from "@cliffy/command";
import { omit } from "@std/collections/omit";
import { expandGlob } from "@std/fs";
import { join } from "@std/path";
import { OPTIONAL_FIELDS, VARIANT_FIELDS } from "./fields.ts";
import { copyExifToVariants, getPhoto } from "./photo.ts";
import type { Photo } from "./types.ts";

const SOURCE_FILE = "source.jpg";

async function isFile(path: string): Promise<boolean> {
  try {
    const info = await Deno.lstat(path);
    return info.isFile;
  } catch {
    return false;
  }
}

async function* getPhotos(photos: string[]): AsyncGenerator<Photo> {
  if (photos.length === 0) {
    for await (const entry of expandGlob(join("*", SOURCE_FILE))) {
      yield await getPhoto(entry.path);
    }
  }
  for (const photo of photos) {
    if (await isFile(photo)) {
      yield (await getPhoto(photo));
    } else if (await isFile(join(photo, SOURCE_FILE))) {
      yield getPhoto(join(photo, SOURCE_FILE));
    }
  }
}

function check(data: Photo) {
  const result = [];
  for (const [field, value] of Object.entries(omit(data, OPTIONAL_FIELDS))) {
    if (!value) result.push(`missing:${field}`);
  }
  for (const size of data.variants) {
    for (const field in omit(size, VARIANT_FIELDS)) {
      if (field in data) result.push(`${size.src}:${field}`);
    }
  }
  if (result.length) return `[${colors.yellow(result.join(", "))}]`;
  return "";
}

function getCommand() {
  return new Command()
    .name("photos")
    .example("photos", "Lists all photos under current directory.")
    .example("photos [photo] --json", "Data for a photo with all variants.")
    .example("photos [photo] --copy", "Copy EXIF data to all variants.")
    .description("Manage photos.")
    .arguments("[photos...:file]")
    .option("--copy", "Copy the EXIF from source JPG to other variants.")
    .option("--json", "Output the EXIF information as JSON.")
    .action(async ({ copy, json }, ...photos) => {
      for await (let photo of getPhotos(photos)) {
        if (copy) {
          await copyExifToVariants(photo);
          photo = await getPhoto(photo.src);
        }
        if (json) console.log(JSON.stringify(photo));
        else console.log(`ðŸ–¼  ${photo.title} ${check(photo)}`);
      }
    });
}

/** CLI entrypoint. */
export async function main(): Promise<void> {
  const command = getCommand();
  await command.parse();
}
