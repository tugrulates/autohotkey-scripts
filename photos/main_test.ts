import { copy } from "@std/fs";
import { join } from "@std/path";
import { assertSnapshot } from "@std/testing/snapshot";
import { mockConsole } from "@tugrulates/testing";
import { main } from "./main.ts";

Deno.test(
  "photos --help",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    using console = mockConsole();
    await main(["--help"]);
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "photos [photo]",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    using console = mockConsole();
    await main(["photos/testdata/floating-around"]);
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "photos [photos]",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    using console = mockConsole();
    await main([
      "photos/testdata/floating-around",
      "photos/testdata/winter-pause",
    ]);
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "photos [file]",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    using console = mockConsole();
    await main(["photos/testdata/floating-around/source.jpg"]);
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "photos [photo] --json",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    using console = mockConsole();
    await main(["photos/testdata/floating-around", "--json"]);
    await assertSnapshot(t, console.calls);
  },
);

Deno.test(
  "photos [photo] --copy",
  { sanitizeOps: false, sanitizeResources: false },
  async (t) => {
    const dir = await Deno.makeTempDir();
    try {
      await copy("photos/testdata", dir, { overwrite: true });
      const photo = join(dir, "winter-pause");
      await t.step("before", async () => {
        using console = mockConsole();
        await main([photo]);
        await assertSnapshot(t, console.calls);
      });

      await t.step("copy", async () => {
        using console = mockConsole();
        await main([photo, "--copy"]);
        await assertSnapshot(t, console.calls);
      });

      await t.step("after", async () => {
        using console = mockConsole();
        await main([photo]);
        await assertSnapshot(t, console.calls);
      });
    } finally {
      await Deno.remove(dir, { recursive: true });
    }
  },
);
