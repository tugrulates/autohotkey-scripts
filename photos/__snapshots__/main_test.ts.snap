export const snapshot = {};

snapshot[`photos --help 1`] = `
[
  {
    data: [
      "
Usage: photos [photos...]

Description:

  Manage photos.

Options:

  -h, --help  - Show this help.                                   
  --copy      - Copy the EXIF from source JPG to other variants.  
  --json      - Output the EXIF information as JSON.              

Examples:

  photos                Lists all photos under current directory.
  photos [photo] --json Data for a photo with all variants.      
  photos [photo] --copy Copy EXIF data to all variants.          
",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photo] 1`] = `
[
  {
    data: [
      "ðŸ–¼  Floating Around",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photos] 1`] = `
[
  {
    data: [
      "ðŸ–¼  Floating Around",
    ],
    level: "log",
  },
  {
    data: [
      "ðŸ–¼  Winter Pause [%cpost-1.jpg:title post-1.jpg:keywords post-1.jpg:date post-1.jpg:location post-1.jpg:city post-1.jpg:state post-1.jpg:country post-1.jpg:camera post-1.jpg:lens post-1.jpg:software post-1.jpg:license post-2.jpg:title post-2.jpg:keywords post-2.jpg:date post-2.jpg:location post-2.jpg:city post-2.jpg:state post-2.jpg:country post-2.jpg:camera post-2.jpg:lens post-2.jpg:software post-2.jpg:license square.jpg:title square.jpg:keywords square.jpg:date square.jpg:location square.jpg:city square.jpg:state square.jpg:country square.jpg:camera square.jpg:lens square.jpg:software square.jpg:license wide.jpg:title wide.jpg:keywords wide.jpg:date wide.jpg:location wide.jpg:city wide.jpg:state wide.jpg:country wide.jpg:camera wide.jpg:lens wide.jpg:software wide.jpg:license%c]",
      "color: yellow",
      "color: reset",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [file] 1`] = `
[
  {
    data: [
      "ðŸ–¼  Floating Around",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photo] --json 1`] = `
[
  {
    data: [
      '{"id":"floating-around","src":"photos/testdata/floating-around/source.jpg","width":240,"height":180,"title":"Floating Around","description":"Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting.","keywords":["California Academy Of Sciences","Translucent Bodies","Long Tentacles","Float Gracefully","Vivid Background","Aquarium Setting","Marine Life","Underwater World","Ocean Creatures","Exotic Jellyfish","Aquatic Environment","Sea Life","Ethereal Beauty","Marine Biology","Underwater Exploration","Aquatic Creatures","Jellyfish Swimming","Oceanic Wildlife","Marine Exhibit","Horizontal","No People","Photography","Jellyfish","Underwater","Water","Blue","Undersea","Sea","Nature","Animal Themes","Animal Wildlife","Swimming","Tentacle"],"date":"2023-12-10T14:28:09-08:00","location":"California Academy of Sciences","city":"San Francisco","state":"California","country":"United States","camera":"Apple iPhone 13 mini","lens":"iPhone 13 mini back dual wide camera 5.1mm f/1.6","software":"Affinity Photo 2 2.5.7","license":"https://creativecommons.org/licenses/by/4.0/","variants":[{"src":"photos/testdata/floating-around/post-1.jpg","width":108,"height":135,"description":"Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting (image 1)."},{"src":"photos/testdata/floating-around/post-2.jpg","width":108,"height":135,"description":"Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting (image 2)."},{"src":"photos/testdata/floating-around/square.jpg","width":180,"height":180},{"src":"photos/testdata/floating-around/story.jpg","width":108,"height":192},{"src":"photos/testdata/floating-around/wide.jpg","width":320,"height":180}]}',
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photo] --copy 1`] = `
[
  {
    data: [
      "ðŸ–¼  Winter Pause [%cpost-1.jpg:title post-1.jpg:keywords post-1.jpg:date post-1.jpg:location post-1.jpg:city post-1.jpg:state post-1.jpg:country post-1.jpg:camera post-1.jpg:lens post-1.jpg:software post-1.jpg:license post-2.jpg:title post-2.jpg:keywords post-2.jpg:date post-2.jpg:location post-2.jpg:city post-2.jpg:state post-2.jpg:country post-2.jpg:camera post-2.jpg:lens post-2.jpg:software post-2.jpg:license square.jpg:title square.jpg:keywords square.jpg:date square.jpg:location square.jpg:city square.jpg:state square.jpg:country square.jpg:camera square.jpg:lens square.jpg:software square.jpg:license wide.jpg:title wide.jpg:keywords wide.jpg:date wide.jpg:location wide.jpg:city wide.jpg:state wide.jpg:country wide.jpg:camera wide.jpg:lens wide.jpg:software wide.jpg:license%c]",
      "color: yellow",
      "color: reset",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photo] --copy 2`] = `
[
  {
    data: [
      "ðŸ–¼  Winter Pause",
    ],
    level: "log",
  },
]
`;

snapshot[`photos [photo] --copy 3`] = `
[
  {
    data: [
      "ðŸ–¼  Winter Pause",
    ],
    level: "log",
  },
]
`;
